crispr_read_counts/single_guide_count.py

```diff
@@ -0,0 +1,249 @@
+            if count == 0:
+              zero_count_guides += 1
+            if count < self.LOW_COUNT_GUIDES_THRESHOLD:
+              low_count_guides += 1
+            f.write('\t'.join([sgrna_id, self.targeted_genes[sgrna_id], str(count)]) + '\n')
+
+    self.stats['zero_count_guides'] = zero_count_guides
+    self.stats['low_count_guides'] = low_count_guides
+    with open(out_stats, 'w') as out_s:
+      json.dump(self.stats, out_s)
+      out_s.write('\n')
+
+  def _write_output_without_stats(self):
+    with open(self.out_count, 'w') as f:
+      if self.plas_name:
+        f.write('\t'.join(['sgRNA', 'gene', f'{self.sample_name}.sample', self.plas_name]) + '\n')
+        for sgrna_seq in (sorted(self.lib.keys())):
+          for sgrna_id in self.lib[sgrna_seq]:
+            count = self.sample_count.get(sgrna_id, 0)
+            plasmid_count = self.plasmid.get(sgrna_id, 0)
+            f.write('\t'.join([sgrna_id, self.targeted_genes[sgrna_id], str(count), str(plasmid_count)]) + '\n')
+      else:
+        f.write('\t'.join(['sgRNA', 'gene', f'{self.sample_name}.sample']) + '\n')
+        for sgrna_seq in (sorted(self.lib.keys())):
+          for sgrna_id in self.lib[sgrna_seq]:
+            count = self.sample_count.get(sgrna_id, 0)
+            f.write('\t'.join([sgrna_id, self.targeted_genes[sgrna_id], str(count)]) + '\n')
+
+  def count(self, trim, plasmid_count_file, reverse_complement, out_stats):
+    if plasmid_count_file:
+      self.plasmid, self.plas_name = self.get_plasmid_read_counts(plasmid_count_file)
+    self.get_sgrna_library_counts(trim, reverse_complement, bool(out_stats))
+    self.write_output(out_stats)
+
+  @staticmethod
+  def get_single_guide_library(lib_file: str):
+    lib = {}
+    targeted_genes = {}
+
+    with open(lib_file, 'r') as f:
+      for line_count, line in enumerate(f):
+        ele = line.strip().split(',')
+        if len(ele) < 3:
+          sys.exit(error_msg(f'Guide RNA library file line: {line_count + 1} does not have 3 columns, or the file uses execpted delimiter.'))
+        sgrna_id, gene_name, lib_seq = ele[0], ele[1], ele[2]
+        if not DNA_PATTERN.match(lib_seq):
+          sys.exit(error_msg(f'Sequence column contains non-DNA characters on line: {line_count}.'))
+
+        if lib_seq in lib:
+          lib[lib_seq].append(sgrna_id)
+        else:
+          lib[lib_seq] = [sgrna_id]
+
+        targeted_genes[sgrna_id] = gene_name
+
+    return lib, targeted_genes
+
+  @staticmethod
+  def get_plasmid_read_counts(plasmid_file: str):
+    plasmid, plasmid_name = {}, {}
+    with open(plasmid_file, 'r') as f:
+      for line_count, line in enumerate(f):
+        ele = line.strip().split('\t')
+        if len(ele) < 3:
+          sys.exit(error_msg(f'Plasmid count file line: {line_count + 1} does not have 3 columns, or the file uses execpted delimiter.'))
+        if PLASMID_COUNT_HEADER.match(line):
+          plasmid_name = ele[2]
```

Comment:
The `plasmid_name` was initialised as a dictionary but here it is assigned a string and may be overwritten multiple times.